src/zjs_web_sockets.c

```diff
@@ -0,0 +1,802 @@
+    ZVAL data_buf = zjs_buffer_create(len, &zbuf);
+    // copy data from read buffer
+    memcpy(zbuf->buffer, con->rptr, len);
+    con->rptr += len;
+    if (con->rptr == con->wptr) {
+        con->rptr = con->wptr = con->rbuf;
+        DBG_PRINT("all data consumed, reseting read buffer\n");
+    }
+    switch (opcode) {
+    case WS_PACKET_CONTINUATION:
+        // continuation frame
+    case WS_PACKET_TEXT_DATA:
+        // text data
+        zjs_trigger_event(con->conn, "message", &data_buf, 1, NULL, NULL);
+        break;
+    case WS_PACKET_BINARY_DATA:
+        // binary data
+        break;
+    case WS_PACKET_PING:
+        zjs_trigger_event(con->conn, "ping", &data_buf, 1, NULL, NULL);
+        break;
+    case WS_PACKET_PONG:
+        zjs_trigger_event(con->conn, "pong", &data_buf, 1, NULL, NULL);
+        break;
+    default:
+        ERR_PRINT("opcode 0x%02x not recognized\n", opcode);
+        break;
+    }
+}
+
+static void close_connection(ws_connection_t *con)
+{
+    ws_connection_t *cur = connections;
+    if (cur->next == NULL) {
+        if (cur != con) {
+            ERR_PRINT("connection handle %p does not exist\n", con);
+            return;
+        }
+        connections = NULL;
+        net_context_put(cur->tcp_sock);
+        zjs_free(cur->rbuf);
+        if (cur->accept_key) {
+            zjs_free(cur->accept_key);
+        }
+        zjs_remove_callback(cur->trigger_msg_id);
+        zjs_remove_callback(cur->accept_handler_id);
+        zjs_free(cur);
+        DBG_PRINT("Freed socket: opened_sockets=%p\n", connections);
+    } else {
+        while (cur->next) {
+            if (cur->next == con) {
+                cur->next = cur->next->next;
+                net_context_put(cur->tcp_sock);
+                zjs_free(cur->rbuf);
+                if (cur->accept_key) {
+                    zjs_free(cur->accept_key);
+                }
+                zjs_remove_callback(cur->trigger_msg_id);
+                zjs_remove_callback(cur->accept_handler_id);
+                zjs_free(cur);
+                DBG_PRINT("Freed socket: opened_sockets=%p\n", connections);
+            }
+        }
+    }
+}
+
+// process a TCP WS packet
+static void process_packet(ws_connection_t *con, uint8_t *data, uint32_t len)
+{
+    ws_packet_t *packet = zjs_malloc(sizeof(ws_packet_t));
+    if (!packet) {
+        ERR_PRINT("allocation failed\n");
+        ZVAL_MUTABLE error = zjs_custom_error("OutOfMemoryError", "could not allocate packet");
+        jerry_value_clear_error_flag(&error);
+        zjs_trigger_event(con->conn, "error", &error, 1, NULL, NULL);
+        return;
+    }
+    memset(packet, 0, sizeof(ws_packet_t));
+    if (decode_packet(packet, data, len) < 0) {
+        ERR_PRINT("error decoding packet\n");
+        zjs_free(packet->payload);
+        zjs_free(packet);
+        ZVAL_MUTABLE error = zjs_custom_error("PacketDecodeError", "error decoding packet");
+        jerry_value_clear_error_flag(&error);
```

Comment:
because you cant use a object that has the error flag set as a parameter. Its a JrS limitation.